ethereum/contracts/pyth/PythGovernance.sol

```diff
@@ -0,0 +1,115 @@
+ */
+abstract contract PythGovernance is Pyth, PythGovernanceInstructions {
+    event ContractUpgraded(address oldContract, address newContract);
+    event GovernanceDataSourceSet(PythInternalStructs.DataSource oldDataSource, PythInternalStructs.DataSource newDataSource);
+    event DataSourcesSet(PythInternalStructs.DataSource[] oldDataSources, PythInternalStructs.DataSource[] newDataSources);
+    event FeeSet(uint oldFee, uint newFee);
+    event ValidPeriodSet(uint oldValidPeriod, uint newValidPeriod);
+
+    function verifyGovernanceVM(bytes memory encodedVM) internal returns (IWormhole.VM memory parsedVM){
+        (IWormhole.VM memory vm, bool valid, string memory reason) = wormhole().parseAndVerifyVM(encodedVM);
+
+        require(valid, reason);
+        require(isValidGovernanceDataSource(vm.emitterChainId, vm.emitterAddress), "VAA is not coming from the governance data source");
+
+        require(vm.sequence > lastExecutedGovernanceSequence(), "VAA is older than the last executed governance VAA");
+
+        setLastExecutedGovernanceSequence(vm.sequence);
+
+        return vm;
+    }
+
+    function executeGovernanceInstruction(bytes calldata encodedVM) public {
+        IWormhole.VM memory vm = verifyGovernanceVM(encodedVM);
+
+        GovernanceInstruction memory gi = parseGovernanceInstruction(vm.payload);
+
+        require(gi.targetChainId == chainId() || gi.targetChainId == 0, "Invalid target chain for this governance instruction");
+
+        // We are explicitly checking with number as with enum there might be confusions
+        // about the numbers.
+        if (gi.action == 1) { // UpgradeContract
+            require(gi.targetChainId != 0, "Upgrade for all the chains does not exists");
+            upgradeContract(gi.payload);
+        } else if (gi.action == 2) {
+            setGovernanceDataSource(gi.payload);
+        } else if (gi.action == 3) {
+            setDataSources(gi.payload);
+        } else if (gi.action == 4) {
+            setFee(gi.payload);
+        } else if (gi.action == 5) {
+            setValidPeriod(gi.payload);
+        } else {
+            revert("Invalid governance action");
+        }
+    }
+
+    // Execute a UpgradeContract governance message
+    function upgradeContract(bytes memory encodedPayload) internal {
+        UpgradeContractPayload memory payload = parseUpgradeContractPayload(encodedPayload); 
+        upgradeUpgradableContract(payload);
+    }
+
+    function upgradeUpgradableContract(UpgradeContractPayload memory payload) virtual internal;
+
+
+    function setGovernanceDataSource(bytes memory encodedPayload) internal {
+        SetGovernanceDataSourcePayload memory payload = parseSetGovernanceDataSourcePayload(encodedPayload);
+
+        PythInternalStructs.DataSource memory oldGovernanceDatSource = governanceDataSource();
+
+        setGovernanceDataSource(payload.newGovernanceDataSource);
+        setLastExecutedGovernanceSequence(0);
```

Comment:
Resetting the sequence number here might make us vulnerable to someone submitting an older governance action, if we ever change the data source back to a previously used one. I wonder if it would be safer to store sequence numbers per data source?

This is a very unlikely scenario, but maybe worth thinking about.

Reply:
nice catch, I will add 'initialSequenceNumber` to this command.